File: aws/ses/typescript/src/services/account.ts
/**
 * Account Service
 *
 * Service for managing account-level settings and quotas in AWS SES v2.
 *
 * @module services/account
 */

import { BaseService } from './base.js';
import type {
  GetAccountRequest,
  PutAccountDetailsRequest,
  PutAccountSuppressionAttributesRequest,
  GetAccountSuppressionAttributesRequest,
  PutAccountSendingAttributesRequest,
  PutAccountVdmAttributesRequest,
  GetAccountVdmAttributesRequest,
} from '../types/requests.js';
import type {
  GetAccountResponse,
  PutAccountDetailsResponse,
  PutAccountSuppressionAttributesResponse,
  GetAccountSuppressionAttributesResponse,
  PutAccountSendingAttributesResponse,
  PutAccountVdmAttributesResponse,
  GetAccountVdmAttributesResponse,
} from '../types/responses.js';

/**
 * Account service for managing SES account settings.
 *
 * Provides methods for:
 * - Retrieving account details and quotas
 * - Updating account information
 * - Managing account-level suppression settings
 * - Controlling sending enable/disable
 * - Managing Virtual Deliverability Manager (VDM) settings
 *
 * @example
 * ```typescript
 * const accountService = new AccountService(client);
 *
 * // Get account details
 * const account = await accountService.getAccount();
 * console.log('Production access:', account.productionAccessEnabled);
 * console.log('Daily send quota:', account.sendQuota?.max24HourSend);
 * console.log('Max send rate:', account.sendQuota?.maxSendRate);
 * console.log('Sent today:', account.sendQuota?.sentLast24Hours);
 *
 * // Check if in sandbox
 * if (!account.productionAccessEnabled) {
 *   console.log('Account is in sandbox mode');
 * }
 * ```
 */
export class AccountService extends BaseService {
  /**
   * Get account details.
   *
   * Retrieves information about your SES account including:
   * - Production access status
   * - Sending quotas (daily limit, rate limit, current usage)
   * - Sending status (enabled/paused)
   * - Account details (mail type, contact info, etc.)
   *
   * @returns Promise resolving to account details
   *
   * @example
   * ```typescript
   * const account = await accountService.getAccount();
   *
   * // Check sending quota
   * const quota = account.sendQuota;
   * if (quota) {
   *   console.log(`Daily quota: ${quota.sentLast24Hours}/${quota.max24HourSend}`);
   *   console.log(`Max send rate: ${quota.maxSendRate} emails/second`);
   *
   *   const remaining = quota.max24HourSend! - quota.sentLast24Hours!;
   *   console.log(`Remaining today: ${remaining} emails`);
   * }
   *
   * // Check production access
   * if (!account.productionAccessEnabled) {
   *   console.log('⚠️ Sandbox mode - can only send to verified addresses');
   * }
   *
   * // Check sending status
   * console.log('Sending enabled:', account.sendingEnabled);
   *
   * // View account details
   * if (account.details) {
   *   console.log('Mail type:', account.details.mailType);
   *   console.log('Website:', account.details.websiteURL);
   *   console.log('Use case:', account.details.useCaseDescription);
   * }
   * ```
   */
  async getAccount(): Promise<GetAccountResponse> {
    return this.get<GetAccountResponse>('/v2/email/account');
  }

  /**
   * Update account details.
   *
   * Updates information about your use case and business to help AWS
   * review your account for production access.
   *
   * @param request - Account details update request
   * @returns Promise resolving when account details are updated
   *
   * @example Request production access
   * ```typescript
   * await accountService.putAccountDetails({
   *   mailType: 'TRANSACTIONAL',
   *   websiteURL: 'https://example.com',
   *   contactLanguage: 'EN',
   *   useCaseDescription: `
   *     We send transactional emails including:
   *     - Order confirmations
   *     - Shipping notifications
   *     - Password reset emails
   *     Expected volume: 10,000 emails/day
   *   `,
   *   additionalContactEmailAddresses: [
   *     'admin@example.com',
   *     'support@example.com'
   *   ],
   *   productionAccessEnabled: true
   * });
   * ```
   *
   * @example Update for marketing emails
   * ```typescript
   * await accountService.putAccountDetails({
   *   mailType: 'MARKETING',
   *   websiteURL: 'https://example.com',
   *   contactLanguage: 'EN',
   *   useCaseDescription: `
   *     We send marketing newsletters to opted-in subscribers.
   *     - Weekly newsletter
   *     - Product announcements
   *     - All recipients have explicitly subscribed
   *     Expected volume: 50,000 emails/week
   *   `
   * });
   * ```
   */
  async putAccountDetails(request: PutAccountDetailsRequest): Promise<PutAccountDetailsResponse> {
    return this.put<PutAccountDetailsResponse>('/v2/email/account/details', request);
  }

  /**
   * Update account-level suppression settings.
   *
   * Configures which types of addresses are automatically added to the
   * account suppression list (bounces and/or complaints).
   *
   * @param suppressedReasons - Array of suppression reasons to enable
   * @returns Promise resolving when suppression settings are updated
   *
   * @example Enable both bounce and complaint suppression
   * ```typescript
   * await accountService.putAccountSuppressionAttributes(['BOUNCE', 'COMPLAINT']);
   * ```
   *
   * @example Only suppress bounces
   * ```typescript
   * await accountService.putAccountSuppressionAttributes(['BOUNCE']);
   * ```
   */
  async putAccountSuppressionAttributes(
    suppressedReasons: Array<'BOUNCE' | 'COMPLAINT'>
  ): Promise<PutAccountSuppressionAttributesResponse> {
    return this.put<PutAccountSuppressionAttributesResponse>(
      '/v2/email/account/suppression',
      { suppressedReasons }
    );
  }

  /**
   * Get account suppression settings.
   *
   * Retrieves the current account-level suppression configuration.
   *
   * @returns Promise resolving to suppression settings
   *
   * @example
   * ```typescript
   * const response = await accountService.getAccountSuppressionAttributes();
   *
   * console.log('Suppression enabled for:', response.suppressionAttributes?.suppressedReasons);
   * // Example output: ['BOUNCE', 'COMPLAINT']
   * ```
   */
  async getAccountSuppressionAttributes(): Promise<GetAccountSuppressionAttributesResponse> {
    return this.get<GetAccountSuppressionAttributesResponse>('/v2/email/account/suppression');
  }

  /**
   * Update account sending status.
   *
   * Enables or disables email sending for the entire account.
   * When disabled, no emails can be sent from any identity.
   *
   * @param sendingEnabled - Whether to enable sending
   * @returns Promise resolving when sending status is updated
   *
   * @example Pause all sending
   * ```typescript
   * // Temporarily pause all email sending
   * await accountService.putAccountSendingAttributes(false);
   *
   * console.log('All email sending is now paused');
   * ```
   *
   * @example Resume sending
   * ```typescript
   * // Resume email sending
   * await accountService.putAccountSendingAttributes(true);
   *
   * console.log('Email sending is now enabled');
   * ```
   */
  async putAccountSendingAttributes(sendingEnabled: boolean): Promise<PutAccountSendingAttributesResponse> {
    return this.put<PutAccountSendingAttributesResponse>(
      '/v2/email/account/sending',
      { sendingEnabled }
    );
  }

  /**
   * Update Virtual Deliverability Manager (VDM) settings.
   *
   * Configures VDM features including dashboard and engagement tracking.
   * VDM provides insights and recommendations to improve deliverability.
   *
   * @param request - VDM settings request
   * @returns Promise resolving when VDM settings are updated
   *
   * @example Enable VDM features
   * ```typescript
   * await accountService.putAccountVdmAttributes({
   *   vdmAttributes: {
   *     dashboardOptions: {
   *       engagementMetrics: 'ENABLED'
   *     },
   *     guardianOptions: {
   *       optimizedSharedDelivery: 'ENABLED'
   *     }
   *   }
   * });
   * ```
   *
   * @example Disable VDM
   * ```typescript
   * await accountService.putAccountVdmAttributes({
   *   vdmAttributes: {
   *     dashboardOptions: {
   *       engagementMetrics: 'DISABLED'
   *     },
   *     guardianOptions: {
   *       optimizedSharedDelivery: 'DISABLED'
   *     }
   *   }
   * });
   * ```
   */
  async putAccountVdmAttributes(request: PutAccountVdmAttributesRequest): Promise<PutAccountVdmAttributesResponse> {
    return this.put<PutAccountVdmAttributesResponse>('/v2/email/account/vdm', request);
  }

  /**
   * Get Virtual Deliverability Manager (VDM) settings.
   *
   * Retrieves the current VDM configuration for the account.
   *
   * @returns Promise resolving to VDM settings
   *
   * @example
   * ```typescript
   * const response = await accountService.getAccountVdmAttributes();
   *
   * const vdm = response.vdmAttributes;
   * console.log('Dashboard enabled:', vdm?.dashboardOptions?.engagementMetrics === 'ENABLED');
   * console.log('Guardian enabled:', vdm?.guardianOptions?.optimizedSharedDelivery === 'ENABLED');
   * ```
   */
  async getAccountVdmAttributes(): Promise<GetAccountVdmAttributesResponse> {
    return this.get<GetAccountVdmAttributesResponse>('/v2/email/account/vdm');
  }
}

File: aws/ses/rust/src/services/account.rs
//! Account-level operations for SES v2.
//!
//! This module provides methods for querying and updating account-level
//! settings and quotas.

use std::sync::Arc;
use serde::{Deserialize, Serialize};

use crate::error::{SesError, SesResult};
use crate::http::{HttpClient, SesRequest, HttpMethod};
use crate::types::SuppressionOptions;
use super::SesService;

/// Service for account-level operations.
///
/// This service provides methods for:
/// - Getting account information and quotas
/// - Updating account details
/// - Configuring account-level sending attributes
/// - Configuring account-level suppression attributes
pub struct AccountService {
    http_client: Arc<dyn HttpClient>,
}

impl AccountService {
    /// Create a new account service.
    pub fn new(http_client: impl HttpClient + 'static) -> Self {
        Self {
            http_client: Arc::new(http_client),
        }
    }

    /// Get information about the account.
    ///
    /// # Returns
    ///
    /// Account details including sending quota, production access status,
    /// and enforcement status.
    pub async fn get_account(&self) -> SesResult<GetAccountResponse> {
        let ses_request = SesRequest::new(HttpMethod::Get, "/v2/email/account");
        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize GetAccount response: {}", e),
            })
    }

    /// Update account details.
    ///
    /// # Arguments
    ///
    /// * `production_access_enabled` - Whether production access is enabled
    /// * `mail_type` - Type of mail being sent
    /// * `website_url` - Website URL
    /// * `contact_language` - Preferred contact language
    /// * `use_case_description` - Description of the use case
    pub async fn put_account_details(
        &self,
        production_access_enabled: Option<bool>,
        mail_type: Option<String>,
        website_url: Option<String>,
        contact_language: Option<String>,
        use_case_description: Option<String>,
    ) -> SesResult<PutAccountDetailsResponse> {
        let request = PutAccountDetailsRequest {
            production_access_enabled,
            mail_type,
            website_url,
            contact_language,
            use_case_description,
            additional_contact_email_addresses: None,
        };

        let body = serde_json::to_vec(&request)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize PutAccountDetails request: {}", e),
            })?;

        let ses_request = SesRequest::new(HttpMethod::Put, "/v2/email/account/details")
            .with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize PutAccountDetails response: {}", e),
            })
    }

    /// Update account sending attributes.
    ///
    /// # Arguments
    ///
    /// * `sending_enabled` - Whether sending is enabled for the account
    pub async fn put_account_sending_attributes(
        &self,
        sending_enabled: Option<bool>,
    ) -> SesResult<PutAccountSendingAttributesResponse> {
        let request = PutAccountSendingAttributesRequest { sending_enabled };

        let body = serde_json::to_vec(&request)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize PutAccountSendingAttributes request: {}", e),
            })?;

        let ses_request = SesRequest::new(HttpMethod::Put, "/v2/email/account/sending")
            .with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize PutAccountSendingAttributes response: {}", e),
            })
    }

    /// Update account suppression attributes.
    ///
    /// # Arguments
    ///
    /// * `suppression_options` - Suppression options to set
    pub async fn put_account_suppression_attributes(
        &self,
        suppression_options: SuppressionOptions,
    ) -> SesResult<PutAccountSuppressionAttributesResponse> {
        let body = serde_json::to_vec(&suppression_options)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize suppression options: {}", e),
            })?;

        let ses_request = SesRequest::new(HttpMethod::Put, "/v2/email/account/suppression")
            .with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize response: {}", e),
            })
    }
}

impl SesService for AccountService {
    fn http_client(&self) -> &Arc<dyn HttpClient> {
        &self.http_client
    }
}

// Request types

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct PutAccountDetailsRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    production_access_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    mail_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    website_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    contact_language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    use_case_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    additional_contact_email_addresses: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct PutAccountSendingAttributesRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    sending_enabled: Option<bool>,
}

// Response types

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct GetAccountResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sending_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dedicated_ip_auto_warmup_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enforcement_status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub production_access_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub send_quota: Option<SendQuota>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub suppression_attributes: Option<SuppressionOptions>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct SendQuota {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_send_rate: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max24_hour_send: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sent_last24_hours: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutAccountDetailsResponse {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutAccountSendingAttributesResponse {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct PutAccountSuppressionAttributesResponse {}
